---
title: "DGEA in CDX expressing ATOH1 and comparison with ATOH1 direct targets from BETA"
author: "Alessia Catozzi"
date: "18/02/2021"
params:
   CDX_folder: "/data/cep/CDX_Model_NGS_Data/RNA_Seq/"
   Nextflow_output_RNASeq: "/data/cep/CDX_Model_NGS_Data/ExVivo_Experiments/ATOH1_KD_ACatozzi/NFpipeline_RNAseq/results_human/"
   Nextflow_output_ChIPSeq: "/data/cep/CDX_Model_NGS_Data/ExVivo_Experiments/ATOH1_KD_ACatozzi/NFpipeline_ChIP/results_human/"
   Design_files: "/data/cep/acatozzi/ATOH1_ChIPSeq/alessias-chip-analysis/Data/"
   Output_data_ATOH1_KD_RNAseq: "/data/cep/acatozzi/ATOH1_ChIPSeq/Output_data/ATOH1_KD_RNAseq/"
   Output_data_ATOH1_ChIP: "/data/cep/acatozzi/ATOH1_ChIPSeq/Output_data/ATOH1_ChIPseq/"
   Output_data_CDX_Bulk: "/data/cep/acatozzi/ATOH1_ChIPSeq/Output_data/CDX_bulk_RNAseq/"
   Output_data_BETA: "/data/cep/acatozzi/ATOH1_ChIPSeq/Output_data/Integrative_analysis/BETA/"
   RDS_output: "/data/cep/acatozzi/ATOH1_ChIPSeq/Output_data/RDSfiles/"
   Output_figures_ATOH1_KD_RNAseq: "/data/cep/acatozzi/ATOH1_ChIPSeq/Output_figures/ATOH1_KD_RNAseq/"
   Output_figures_ATOH1_ChIP: "/data/cep/acatozzi/ATOH1_ChIPSeq/Output_figures/ATOH1_ChIPseq/"
   Output_figures_CDX_Bulk: "/data/cep/acatozzi/ATOH1_ChIPSeq/Output_figures/CDX_bulk_RNAseq/"
   Gene_signatures: "/data/cep/acatozzi/gene_signatures/"
   CDXsampleInfo: "/data/cep/CDX_Model_NGS_Data/cdx-model-all-sequencing/Metadata/2104_CDXsampleInformation.csv"
   CDXcountMatrix: "/data/cep/CDX_Model_NGS_Data/RNA_Seq/CDXBiobank_MouseFiltered_GeneCounts_GRCh38v99.tsv"
   CDXdecoderTable: "/data/cep/CDX_Model_NGS_Data/cdx-model-all-sequencing/Metadata/RNAseq_SampleInformation.csv"
   CDXcountsTPM: "/data/cep/CDX_Model_NGS_Data/RNA_Seq/CDXBiobank_MouseFiltered_TPM_GRCh38v99.tsv"
   gtfFile: "/data/cep/CDX_Model_NGS_Data/Reference/Homo_sapiens.GRCh38.99.gtf"
output:
  html_document:
    code_folding: hide
    df_print: paged
    number_sections: yes
    self_contained: yes
    theme: paper
    toc: yes
    toc_float: yes
editor_options:
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	message = FALSE,
	warning = FALSE,
	include = FALSE
)

library(vsn)
library(ggplot2)
library(pheatmap)
library(RColorBrewer)
library(ggrepel)
library(EnhancedVolcano)
library(readxl)
library(org.Hs.eg.db)
library(ggpubr)
library(rtracklayer)
library(enrichplot)
library(fgsea)
library(gprofiler2)
library(tidyverse)
library(DESeq2)

PurOr_palette <- colorRampPalette(c("darkslateblue", "white", "darkorange3"))(n=300)
YlGnBu <- colorRampPalette(c("wheat1", "lightskyblue1", "dodgerblue4"))(n=300)

```

# *Aim of the analysis*

*Aim of the analysis is to perform DGEA on ATOH1 CDX vs all the others and ASCL1 CDX and perform gene set enrichment analysis for target gene sets - including ATOH1 direct targets as detected by BETA.*

# Set up directories

```{r gene annotation}

gr.ann <- rtracklayer::import(params$gtfFile)

geneLookup <- tibble(gene_id = gr.ann$gene_id, gene_name = gr.ann$gene_name, geneBiotype = gr.ann$gene_biotype) %>%
  distinct() %>% 
  column_to_rownames(var = "gene_id")

```

# Load CDX bulk RNAseq count matrix and metadata

The count data table and decoder table was generated by Sam and includes gene name, gene biotype and Ensembl gene ID. Differential gene expression is performed on Ensembl gene IDs to avoid duplicated gene names.
The count matrix is generated by deleting annotation columns (i.e. gene biotype and gene name) and by transforming the ensembl gene ID column to row names of the new count matrix. This format is then suitable for analysis by DESeq2.


```{r message=FALSE, warning=FALSE}

CDX_info <-  read_csv(params$CDXsampleInfo)

CDX_alt_name <- read_xlsx("/data/cep/acatozzi/catozzi_et_al/Input_Data/CDX_alt_names.xlsx") %>%
  dplyr::select(CDX, alt_name) %>%
  dplyr::rename(cdx_id = CDX) %>%
  dplyr::filter(cdx_id %in% CDX_info$cdx_id)

CDX_info <- CDX_info %>%
  right_join(CDX_alt_name, by = "cdx_id")

decoder <- read_csv(params$CDXdecoderTable) %>% 
					filter(status != "embargo", !(cdx_id %in% c("CDX47", "CDX36"))) %>% 
					left_join(CDX_info, by = "cdx_id")

decoder <- decoder %>%
  mutate(ATOH1_status = ifelse(tf_subtype == "ATOH1", "ATOH1", "other"))

# Check to have all replicates

table(decoder$cdx_id)

```


## Load count matrix and check DESeq object

Here I am loading the count matrix generated by Sam and selecting only the models that I am interested in. For DESeq2 to run correcly, I need to check that the sample names in the count matrix (in the columns) and in the metadata are the same (row names in decoder table).


```{r}
# read in the count matrix

raw_counts <- read_tsv(params$CDXcountMatrix)

# Subset the count matirx based on samples of interest and make the genes rownames

countdata <- raw_counts %>% 
            dplyr::select("gene_id", decoder$sample_id) %>%
            column_to_rownames(var = "gene_id")

```


With *DESeqDataSetFromMatrix* I create a DESeq data set linking the count matrix, the metadata associated to the samples and specifiy the design. The design dictates what we will compare. In this case, I want to compare CDX models that express different TFs, therefore the comparison will be designed based on 'TF' in the metadata.

If you want to change the comparison, i.e. for baseline vs progression, you only need to change 'design = ~ TF' to 'design = ~ status'.

Based on your research question, you can decide to filter out genes with low counts, in a more or less stringent way. Here I have filtered out genes for which the sum of the counts across all the samples is less than 500.
The threshold you choose depends on how many samples you have, your research question and how many differentially expressed genes you have by the end of the analysis.

Here I am creating the design based on the batch (pre- / post-Kris for CDX1-10 and 11-47) because the batch has a very strong effect on what the data looks like (see section below).

```{r Create DESeq object}

dds_ATOH1 <- DESeqDataSetFromMatrix( countData = round(countdata),
                               colData = column_to_rownames(decoder, var = "sample_id"),
                               rowData = raw_counts %>% select(gene_id, gene_name) %>% column_to_rownames(var = "gene_id"),
                               design = ~ ATOH1_status)

dds_ATOH1 <- dds_ATOH1[rowSums(counts(dds_ATOH1)) > 10, ]


```

# Batch effect removal

No batch effect found.

```{r}

vsd_ATOH1 <- vst(dds_ATOH1)

pcaData_ATOH1 <- prcomp(t(assay(vsd_ATOH1)), scale = TRUE, center = TRUE)

pcaData_ATOH1

#table(rownames(pcaData.vst$x) == rownames(decoder))

pcaData <- data.frame(pcaData_ATOH1$x, CDX = decoder$alt_name, batch = decoder$batch, run_id = decoder$run_id)

percentVar <- round(100 * pcaData_ATOH1$sdev^2 / sum(pcaData_ATOH1$sdev^2))

ggplot(pcaData, aes(x = PC1, y = PC2, group = CDX, color = batch)) +
    geom_point(size = 3) +
    xlab(paste0("PC1:", percentVar[1], "% variance")) +
    ylab(paste0("PC2:", percentVar[2], "% variance")) +
    geom_hline(yintercept=0, linetype = "dashed") +
    geom_vline(xintercept=0, linetype = "dashed") +
    theme_bw()

```

# PCA plot

I can now look at how different CDX are located in the reduced dimensionality space and at how neurogenic TFs are expressed across the different CDX models.

```{r, fig.width=10, fig.height=7}

pcaData$TF <- decoder$tf_subtype

ggplot(pcaData, aes(x = PC1, y = PC2)) +
    geom_point(aes(colour = pcaData$TF), alpha = 0.8, size = 5.5) +
    geom_point(shape = 21, colour = "black", size = 6) +
    xlab(paste0("PC1:", percentVar[1], "% variance")) +
    ylab(paste0("PC2:", percentVar[2], "% variance")) +
    theme_classic() +
    theme(legend.position = "none") +
    theme(axis.text = element_text(size = 22)) +
    theme(axis.title = element_text(size = 22)) +
    scale_color_manual(values=c("#08519c", #ASCL1
                                "darkgoldenrod1", #ATOH1
                                "#dd1c77", #NEUROD1
                                "#31a354"))  #POU2F3

```

# Quick sanity check before DGEA

At this point I will plot the distance matrix to make sure each replicate aligns with each other. 
*For more extensive sanity checks see RNAseqSanityChecks.Rmd workflow.*

```{r fig.height=17, fig.width=17, include=FALSE}

sampleDists_ATOH1 <- dist(t(assay(vsd_ATOH1)))
sampleDistMatrix_ATOH1 <- as.matrix(sampleDists_ATOH1)

# table(rownames(sampleDistMatrix) == rownames(decoder))

rownames(sampleDistMatrix_ATOH1) <- decoder$alt_name
colnames(sampleDistMatrix_ATOH1) <- NULL

colour_palette.blues <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)

plot.sampleDistHeatmap <- pheatmap::pheatmap(sampleDistMatrix_ATOH1,
                                    clustering_distance_rows = sampleDists_ATOH1,
                                    clustering_distance_cols = sampleDists_ATOH1,
                                    treeheight_row = 0,
                                    treeheight_col = 0,
                                    fontsize = 14,
                                    show_rownames = TRUE,
                                    color = colour_palette.blues)

```



# DGEA ATOH1 vs others

Here I am creating a dds object to perform differential gene expression across CDX expressing ATOH1 vs the others, independently of the fact that they express ASCL1 or NEUROD1.

```{r}

dds_ATOH1$ATOH1_status <- relevel(dds_ATOH1$ATOH1_status, ref = "other")

dds_ATOH1 <- DESeq(dds_ATOH1)

# Check results names to verify what to use as contrast

resultsNames(dds_ATOH1)

resATOH1_vs_Other <- results(dds_ATOH1, contrast = c("ATOH1_status", "ATOH1", "other"))

print(paste0("Example of results from DGEA of ATOH1 vs Other CDX:"))
resATOH1_vs_Other

resATOH1_vs_Other["ENSG00000172238",] #ATOH1
resATOH1_vs_Other["ENSG00000139352",] #ASCL1

```

## Export normalized counts for GSEA

```{r}

norm_counts <- counts(dds_ATOH1, normalized=TRUE)

table(colnames(norm_counts) == decoder$sample_id)

colnames(norm_counts) <- decoder$sample_id

norm_counts

norm_counts <- (dplyr::select(geneLookup, gene_name)) %>%
  merge(norm_counts, by = 0)                

norm_counts$Row.names <- NULL

norm_counts

#write_tsv(norm_counts, file = paste0(params$Output_data_CDX_Bulk, "DESEQ2_ATOH1vOthers_normalized_counts_forGSEA.tsv"))

```


## Volcano plot

```{r, fig.height=10, fig.width=12}

resShrinkATOH1vsOther <- lfcShrink(dds_ATOH1, coef = "ATOH1_status_ATOH1_vs_other", type = "apeglm")

# annotate results for export

resShrinkATOH1vsOther_annot <- as_tibble(rownames_to_column(as.data.frame(resShrinkATOH1vsOther), var = "gene_id")) %>%
  right_join(rownames_to_column(geneLookup, var = "gene_id"), by = "gene_id") %>%
  dplyr::select(gene_name, everything())

resShrinkATOH1vsOther_annot

write_csv(resShrinkATOH1vsOther_annot, file = paste0(params$Output_data_CDX_Bulk, "CDX_bulk_DGEA_ATOH1vOthers.csv"))

volcanoData <- resShrinkATOH1vsOther %>% 
  as.data.frame() %>%
  merge(geneLookup, by = 0) %>%
  column_to_rownames(var = "Row.names")

labels <- volcanoData %>%
  filter(abs(log2FoldChange) > 10) %>% 
  dplyr::select(gene_name)

volcanoplot <- EnhancedVolcano(volcanoData,
    lab = volcanoData$gene_name,
    x = 'log2FoldChange',
    xlab = bquote(~Log[2]~ 'fold change'),
    y = 'padj',
    pCutoff = 0.01,
    FCcutoff = 2,
    selectLab = c(labels$gene_name, "ASCL1"),
    title ="DE genes ATOH1 vs Other")

volcanoplot

ggsave(paste0(params$Output_figures_CDX_Bulk, "ATOH1vOthers_VolcanoPlot_grey.pdf"), volcanoplot, width = 12, height = 8)

```

Note: importantly, ATOH1 and its putative direct target POU4F3 stand out as differentially expressed in the Volcano plot.

## Extract significant DE genes

```{r}

resATOH1vsOther_significant <- subset(resATOH1_vs_Other, padj < 0.05 & abs(log2FoldChange) > 1.5)

resATOH1vsOther_significant

```

## FGSEA

```{r}
library(fgsea)

top50 <- resShrinkATOH1vsOther_annot %>%
  slice_min(log2FoldChange, n = 50)

# fgsea will be performed on genes ranked by their log10 p adjusted, with the sign of log2 fold change to account for upregulation or downregulation

# Note: because genes upregulated in ATOH1 CDX have - value, I have inverted the log2fold change sign, so that gene set enrichment is displayed with a positive curve. Included top50 DE genes in ATOH1 CDX for control.

gseaDat_bulk <- resShrinkATOH1vsOther_annot %>%
            filter(complete.cases(.)) %>% 
            mutate(rank.score = -sign(log2FoldChange)*log10(padj)) %>%
            dplyr::select(gene_name, rank.score) %>%
            na.omit() %>%
            distinct() %>% 
            deframe() %>%
            sort()

# make table for GSEA

out <- resShrinkATOH1vsOther_annot %>%
            filter(complete.cases(.)) %>% 
            mutate(rank.score = sign(log2FoldChange)*log10(padj)) %>%
            dplyr::select(gene_name, rank.score) %>%
            na.omit() %>%
            distinct()

out

#write.table(out, file = paste0(params$Output_data_CDX_Bulk, "GSEA_ATOH1vOthers_CDXBulk_ranked.txt"), sep = "\t", row.names = F, quote = F)

```

Because ATOH1 is included in the list of direct targets, but it is an artifact due to ATOH1 KD, I am removing it from the signature for FGSEA.

```{r}
#table(gseaDat_bulk < 0)

signatures <- read_csv("/mnt/gpfs2/data/cep/acatozzi/gene_signatures/Custom_GSEA.csv")

#downtargets_100K_p001 <- read_delim(file = paste0(params$Output_data_BETA, "ddspeaks_100K_001/basic_100K_001_downtarget.txt"), delim = "\t")

#uptargets_100K_p001 <- read_delim(file = paste0(params$Output_data_BETA, "ddspeaks_100K_001/basic_100K_001_uptarget.txt"), delim = "\t")

downtargets_10K_p001 <- read_delim(file = paste0(params$Output_data_BETA, "ddspeaks_10K_001/basic10K_downtarget.txt"), delim = "\t")

signatures_list <- list(Epithelial_signature = na.omit(signatures$Epithelial_signature),
                        Mesenchymal_signature = na.omit(signatures$Mesenchymal_signature),
                        NE_signature = na.omit(signatures$NE_signature),
                        NonNE_signature = na.omit(signatures$NNE_signature),
                        hair_cells_Cai2015 = na.omit(signatures$Hair_cell_Cai_et_al_2015),
                        ATOH1_targetome_cerebellum = na.omit(signatures$ATOH1_Cerebellum_targetome_Klisch_et_al_2011),
                        ATOH1_OE_iPSC_signature = na.omit(signatures$ATOH1_OE_iPSCs_Ng_et_al_2020),
                        hair_cells_Menendez2020 = na.omit(signatures$hair_cells_Menendez2020),
                        cerebellar_granule_neurons_Menendez2020 = na.omit(signatures$cerebellar_granule_neurons_Menendez2020),
                        merkel_cells_Menendez2020 = na.omit(signatures$merkel_cells_Menendez2020),
                        goblet_cells_Menendez2020 = na.omit(signatures$goblet_cells_Menendez2020),
                        hair_cells_Burns2015 = na.omit(signatures$hair_cells_Burns2015),
                        downtargets_10K = downtargets_10K_p001$GeneSymbol,
                        top50_DE_ATOH1 = top50$gene_name,
                        NE_signature_2021 = na.omit(signatures$NE_signature_Cai2021),
                        NonNE_signature_2021 = na.omit(signatures$NNE_signature_Cai2021))


fgseaRes <- fgsea(pathways = signatures_list, 
                  stats    = gseaDat_bulk,
                  minSize  = 15,
                  maxSize  = 600,
                  eps = 0.0)

fgseaRes %>% arrange(NES)

#signatures.fgsea$leadingEdge[[5]]

plotEnrichment(signatures_list$NE_signature, gseaDat_bulk) + 
                                          labs(title="NE signature")

plotEnrichment(signatures_list$Mesenchymal_signature, gseaDat_bulk) + 
                                          labs(title="Mesenchymal signature")

plotEnrichment(signatures_list$hair_cells_Cai2015, gseaDat_bulk) + 
                                          labs(title="Hair cell signature (Cai et al., 2015)")

plotEnrichment(signatures_list$ATOH1_targetome_cerebellum, gseaDat_bulk) + 
                                          labs(title="ATOH1 targetome in cerebellum (Klisch et al. 2011)")

plotEnrichment(signatures_list$ATOH1_OE_iPSC_signature, gseaDat_bulk) + 
                                          labs(title="ATOH1 OE in iPSCs (Ng et al 2021)")

plotEnrichment(signatures_list$downtargets_10K, gseaDat_bulk) + 
                                          labs(title="ATOH1 direct down-targets 10K")

plotEnrichment(signatures_list$top50_DE_ATOH1, gseaDat_bulk) + 
                                          labs(title="top 50 DE in ATOH1 CDX")

plotEnrichment(signatures_list$merkel_cells_Menendez2020, gseaDat_bulk) + 
                                          labs(title="Merkel cell signature (Menendez et al. 2020")




```


Note: genes enriched in ATOH1 CDX are on the left; overall, genes enriched in ATOH1 CDX (log2 fold change < 0) are 14417 while genes enriched in other CDX are 19634.

```{r}

exp <- fgseaRes %>%
  as.data.frame() %>%
  mutate(leading_edge = sapply(leadingEdge, toString), stringsAsFactors=FALSE)

exp$leadingEdge <- NULL

exp

write_csv(exp, file = paste0(params$Output_data_CDX_Bulk, "fgsea_CDXBulk_ATOH1vOthers_gene_signatures.csv"))

pdf(file = paste0(params$Output_figures_CDX_Bulk, "FGSEA_hair_cell_signature.pdf"), width=9, height=6)

plotEnrichment(signatures_list$hair_cells_Cai2015, gseaDat_bulk) + 
                                          labs(title="Hair cell signature (Cai et al., 2015)")

dev.off()

pdf(file=paste0(params$Output_figures_CDX_Bulk, "FGSEA_cerebellum_signature.pdf"), width=9, height=6)

plotEnrichment(signatures_list$ATOH1_targetome_cerebellum, gseaDat_bulk) + 
                                          labs(title="ATOH1 targetome in cerebellum (Klisch et al. 2011)")

dev.off()


pdf(file=paste0(params$Output_figures_CDX_Bulk, "FGSEA_ATOH1_direct_downregulated_targets_10Kp001.pdf"), width=9, height=6)

plotEnrichment(signatures_list$downtargets_10K, gseaDat_bulk) + 
                                          labs(title="ATOH1 direct down-targets 10K")

dev.off()

pdf(file=paste0(params$Output_figures_CDX_Bulk, "FGSEA_Merkel cell signature_Menendez 2020.pdf"), width=9, height=6)

plotEnrichment(signatures_list$merkel_cells_Menendez2020, gseaDat_bulk) + 
                                          labs(title="Merkel cell signature (Menendez et al. 2020")


dev.off()


```
## Volcano plot of direct targets

```{r, fig.width=30, fig.height=30}

#cap data to -log10(padj) = 100

volcanoData <- resShrinkATOH1vsOther_annot %>%
  mutate(capped_padj = ifelse(-log10(padj) > 100, 1e-100, padj)) %>%
  column_to_rownames(var = "gene_id")

volcanoData %>% arrange(padj)

volcanoData <- volcanoData %>%
  mutate(fill = ifelse(abs(volcanoData$log2FoldChange) > 2 & volcanoData$padj < 0.01, "DE", "other"))

volcanoData <- volcanoData %>%
  mutate(fill = ifelse(volcanoData$gene_name %in% downtargets_10K_p001$GeneSymbol, "direct targets", fill))

volcanoplot_withTargets_bulk <- ggplot(data = volcanoData, aes(x = log2FoldChange, y = -log10(capped_padj), col = fill)) + 
  geom_point(data = volcanoData %>% filter(fill %in% c("DE", "other", NA)), alpha = 0.7, size = 3) +
  geom_point(data = volcanoData %>% filter(fill == "direct targets"), alpha = 1, size = 5) +
  theme_minimal() +
  theme(panel.grid = element_blank(), 
        axis.line = element_line(colour = "black"),
        axis.title = element_text(size = 40), 
        axis.text = element_text(size = 30)) +
  labs(x = bquote(~Log[2]~ 'fold change'), y = "-log10(p adjust)") +
  scale_color_manual(values = c("grey40","red2", "gray60", "gray60")) +
  geom_vline(xintercept=c(-2, 2), col="black", linetype='dashed') +
  geom_hline(yintercept=-log10(0.01), col="black", linetype='dashed') +
  geom_label_repel(data = volcanoData %>% filter(gene_name %in% downtargets_10K_p001$GeneSymbol & log2FoldChange > 0 & -log10(padj) > 3), aes(label = gene_name), color = "black", size = 12, segment.color = NA, check_overlap = T) +
  geom_label_repel(data = volcanoData %>% filter(gene_name == "ASCL1"), aes(label = gene_name), color = "black", size = 12, segment.color = NA, check_overlap = T)

volcanoplot_withTargets_bulk

ggsave(paste0(params$Output_figures_CDX_Bulk, "ATOH1vOthers_VolcanoPlot_withATOH1targets10K_black.pdf"), volcanoplot_withTargets_bulk, width = 12, height = 12)

```

## Heatmap of ATOH1 direct targets

```{r}

sample_names <- subset(decoder, select = c("sample_id", "alt_name")) %>% column_to_rownames(var = "sample_id")

vsd.agg <- merge(sample_names, t(assay(vsd_ATOH1)), by = 0)
vsd.agg$Row.names <- NULL

# Calculate the mean normalized count per CDX model

vsd.agg <- aggregate(vsd.agg[,-1], list(vsd.agg$alt_name), mean)
vsd.agg$alt_name <- NULL
vsd.agg <- column_to_rownames(vsd.agg, var = "Group.1")

vsd.annot <-merge(geneLookup, t(vsd.agg), by = 0) %>%
  dplyr::rename("ensembl_gene_id" = "Row.names") %>%
  dplyr::select("gene_name", "ensembl_gene_id", starts_with("CDX"))

print(paste0("This is what the annotated and aggregated VST normalized counts table looks like:"))
vsd.annot[1:5, 1:5]

# Get column annotation for CDX models, linking each CDX to the TF of interest

colInfo.agg <- dplyr::select(decoder, c("alt_name","tf_subtype"))
colInfo.agg <- colInfo.agg[!duplicated(colInfo.agg$alt_name),]
rownames(colInfo.agg) <- NULL
colInfo.agg <- column_to_rownames(colInfo.agg, var = "alt_name")

```

```{r fig.height=8, fig.width=12}

GOI <- downtargets_10K_p001$GeneSymbol

plotData <- subset(vsd.annot, gene_name %in% GOI)

rownames(plotData) <- NULL

plotData <- dplyr::select(plotData, "ensembl_gene_id", starts_with("CDX")) %>%
  column_to_rownames(var = "ensembl_gene_id")

plotHeatmap <- pheatmap::pheatmap((plotData),
                  color = PurOr_palette,
                  #border_color = "grey60",
                  treeheight_row = 15,
                  treeheight_col = 15,
                  cluster_rows = TRUE,
                  cluster_cols = TRUE,
                  dendrogram = "both",
                  scale = "row",
                  trace = "none",
                  fontsize_row = 14,
                  cellwidth = 12,
                  cellheight = 2,
                  show_rownames = FALSE,
                  annotation_col = colInfo.agg, 
                  annotation_colors = list(TF = c(ASCL1 = "#08519c", NEUROD1 = "#dd1c77", POU2F3 = "#31a354", ATOH1 = "#feb24c")),
                  main = "Heatmap of ATOH1 direct targets in CDX biobank")



```

## GO enirhcment analysis:gProfiler

```{r fig.width=12, fig.height=12}

downregulated <- resShrinkATOH1vsOther_annot %>%
  filter(log2FoldChange < -2 & padj < 0.05) %>%
  arrange(padj) %>%
  dplyr::select(gene_name, padj) %>%
  deframe()

upregulated <- resShrinkATOH1vsOther_annot %>%
  filter(log2FoldChange > 2 & padj < 0.05) %>%
  arrange(padj) %>%
  dplyr::select(gene_name, padj) %>%
  deframe()

# run gprofiler2

gp_down <- gost(names(downregulated), organism = "hsapiens", ordered_query = TRUE, as_short_link = FALSE)

gp_up <- gost(names(upregulated), organism = "hsapiens", ordered_query = TRUE, as_short_link = FALSE)

gp_down$result
gp_up$result

# negative p values for ASCL1, positive p values for ATOH1 CDX

BP <- gp_down$result %>% 
  filter(source == "GO:BP") %>% 
  mutate(p_value = p_value*-1) %>%
  rbind(filter(gp_up$result, gp_up$result$source == "GO:BP")) %>%
  arrange(p_value)

BP$term_name <- make.unique(BP$term_name)

BP

# export GO enrichment analysis as csv

write_csv(BP, file = paste0(params$Output_data_CDX_Bulk, "CDX_Bulk_DGEA_ATOH1vOthers_GProfiler.csv"))


gp.results <- ggplot(data = BP, 
                     aes(x = factor(term_name, levels = pull(arrange(BP, desc(p_value)), term_name)), 
                         y = sign(p_value) * -log10(abs(p_value)))) +
  geom_bar(stat = "identity", colour  = "black", width = .8) + 
  geom_hline(yintercept = -log10(0.05), colour = 'red') +
  geom_hline(yintercept = +log10(0.05), colour = 'red') +
    theme_bw() +
  coord_flip() +
  theme(text = element_text(size = 18),
      	axis.text.y = element_text(size = 14),
        axis.text.x = element_text(size = 18)) +
  labs(y = ("-log10(p value)"), x = "GO Biological Process")
       
gp.results


knitr::knit_exit()

```




# DGEA ATOH1 vs ASCL1

Here I am creating a dds object to perform differential gene expression across CDX expressing ATOH1 vs ASCL1 CDX.

```{r}

dds_ATOH1vASCL1 <- DESeqDataSetFromMatrix( countData = countdata,
                               colData = decoder,
                               rowData = gene_info,
                               design = ~ batch + TF)

print(paste0("Total genes present in the count matrix:", nrow(dds_ATOH1vASCL1)))
dds_ATOH1vASCL1 <- dds_ATOH1vASCL1[rowSums(counts(dds_ATOH1vASCL1)) >= 10,]
print(paste0("Filtered genes with >= 10 reads on which to perform DGEA:", nrow(dds_ATOH1vASCL1)))

dds_ATOH1vASCL1 <- DESeq(dds_ATOH1vASCL1)

# Check results names to verify what to use as contrast

resultsNames(dds_ATOH1vASCL1)

resATOH1_vs_ASCL1 <- results(dds_ATOH1vASCL1, contrast = c("TF",  "ATOH1", "ASCL1"))

print(paste0("Example of results from DGEA of ATOH1 vs ASCL1 CDX:"))
resATOH1_vs_ASCL1

```


## Volcano plot ATOH1 vs ASCL1

```{r, fig.height=6, fig.width=8}

resShrinkATOH1vsASCL1 <- lfcShrink(dds_ATOH1vASCL1, coef = "TF_ATOH1_vs_ASCL1", res = resATOH1_vs_ASCL1, type = "apeglm")

# annotate results for export

resShrinkATOH1vsASCL1_annot <- as_tibble(rownames_to_column(as.data.frame(resShrinkATOH1vsASCL1), var = "gene_id")) %>%
  right_join(rownames_to_column(gene_info, var = "gene_id"), by = "gene_id") %>%
  dplyr::select(gene_id, gene_name, gene_biotype, everything())

write_csv(resShrinkATOH1vsOther_annot, file = paste0(params$Output_data_CDX_Bulk, "CDX_Bulk_DGEA_ASCL1vATOH1.csv"))

volcanoData <- merge(as.data.frame(resShrinkATOH1vsASCL1), gene_info, by = 0)
volcanoData <- column_to_rownames(volcanoData, var = "Row.names")

EnhancedVolcano(volcanoData,
    lab = volcanoData$gene_name,
    x = 'log2FoldChange',
    y = 'padj',
    pCutoff = 0.01,
    FCcutoff = 2,
    #transcriptLabSize = 3.5,
    #transcriptPointSize = 2,
    #transcriptLabvjust = -1.5,
    title ="DE genes ASCL1 vs ATOH1")

```

Note: importantly, ATOH1 and its putative direct target POU4F3 stand out as differentially expressed in the Volcano plot.

## Extract significant DE genes

```{r}

resATOH1vsASCL1_significant <- subset(resATOH1_vs_ASCL1, padj < 0.05 & abs(log2FoldChange) > 1.5)

resATOH1vsASCL1_significant

```

## ATOH1 direct targets FGSEA

```{r}

top50 <- resShrinkATOH1vsASCL1_annot %>%
  slice_max(log2FoldChange, n = 50)

# fgsea will be performed on genes ranked by their log10 p adjusted, with the sign of log2 fold change to account for upregulation or downregulation

# Note: because genes upregulated in ATOH1 CDX have + value, so I have not inverted the log2fold change sign, so that gene set enrichment is displayed with a positive curve. Included top50 DE genes in ATOH1 CDX for control.

gseaDat_ATOH1vASCL1 <- resShrinkATOH1vsASCL1_annot %>%
            filter(complete.cases(.)) %>% 
            mutate(rank.score = sign(-log2FoldChange)*log10(padj)) %>%
            dplyr::select(gene_name, rank.score) %>%
            na.omit() %>%
            distinct() %>% 
            deframe() %>%
            sort()

gseaDat_ATOH1vASCL1

table(gseaDat_ATOH1vASCL1 < 0)

signatures <- read_csv("/mnt/gpfs2/data/cep/acatozzi/gene_signatures/Custom_GSEA.csv")

downtargets_100K_p001 <- read_delim(file = paste0(Output_data_BETA, "ddspeaks_100K_001/basic_100K_001_downtarget.txt"), delim = "\t")

#uptargets_100K_p001 <- read_delim(file = paste0(Output_data_BETA, "ddspeaks_100K_001/basic_100K_001_uptarget.txt"), delim = "\t")
#downtargets_10K_p001 <- read_delim(file = paste0(Output_data_BETA, "ddspeaks_10K_001/basic10K_downtarget.txt"), delim = "\t")

signatures_list <- list(Epithelial_signature = na.omit(signatures$Epithelial_signature),
                        Mesenchymal_signature = na.omit(signatures$Mesenchymal_signature),
                        NE_signature = na.omit(signatures$NE_signature),
                        NonNE_signature = na.omit(signatures$NNE_signature),
                        hair_cells = na.omit(signatures$Hair_cell_Cai_et_al_2015),
                        ATOH1_targetome_cerebellum = na.omit(signatures$ATOH1_Cerebellum_targetome_Klisch_et_al_2011),
                        ATOH1_OE_iPSC_signature = na.omit(signatures$ATOH1_OE_iPSCs_Ng_et_al_2020),
                         downtargets_100K = downtargets_100K_p001$GeneSymbol, 
                        uptargets_100K = uptargets_100K_p001$GeneSymbol,
                        downtargets_10K = downtargets_10K_p001$GeneSymbol,
                        top50_DE_ATOH1 = top50$gene_name)

signatures.fgsea <- fgsea(pathways = signatures_list, stats = gseaDat_ATOH1vASCL1,  minSize=10, maxSize = 4000, nperm = 1000)

signatures.fgsea

#signatures.fgsea$leadingEdge[[5]]

plotEnrichment(signatures_list$NE_signature, gseaDat_ATOH1vASCL1) + 
                                          labs(title="NE signature")

plotEnrichment(signatures_list$NonNE_signature, gseaDat_ATOH1vASCL1) + 
                                          labs(title="NNE signature")

plotEnrichment(signatures_list$Epithelial_signature, gseaDat_ATOH1vASCL1) + 
                                          labs(title="Epithelial signature")

plotEnrichment(signatures_list$Mesenchymal_signature, gseaDat_ATOH1vASCL1) + 
                                          labs(title="Mesenchymal signature")

plotEnrichment(signatures_list$hair_cells, gseaDat_ATOH1vASCL1) + 
                                          labs(title="Hair cell signature (Cai et al., 2015)")

plotEnrichment(signatures_list$ATOH1_targetome_cerebellum, gseaDat_ATOH1vASCL1) + 
                                          labs(title="ATOH1 targetome in cerebellum (Klisch et al. 2011)")

plotEnrichment(signatures_list$ATOH1_OE_iPSC_signature, gseaDat_ATOH1vASCL1) + 
                                          labs(title="ATOH1 OE in iPSCs (Ng et al 2021)")

plotEnrichment(signatures_list$downtargets_100K, gseaDat_ATOH1vASCL1) + 
                                          labs(title="ATOH1 direct targets (BETA, 100K)")

plotEnrichment(signatures_list$uptargets_100K, gseaDat_ATOH1vASCL1) + 
                                          labs(title="ATOH1 direct up targets (BETA, 100K)")

plotEnrichment(signatures_list$downtargets_10K, gseaDat_ATOH1vASCL1) + 
                                          labs(title="ATOH1 direct targets (BETA, 10K)")

plotEnrichment(signatures_list$top50_DE_ATOH1, gseaDat_ATOH1vASCL1) + 
                                          labs(title="top 50 DE in ATOH1 CDX")

exp <- signatures.fgsea %>%
  as.data.frame() %>%
  mutate(leading_edge = sapply(leadingEdge, toString), stringsAsFactors=FALSE)

exp$leadingEdge <- NULL

exp

write_csv(exp, file = paste0(params$Output_data_CDX_Bulk, "fgsea_CDXBulk_ATOH1vASCL1_gene_signatures.csv"))


```
ATOH1 CDX vs ASCL1 CDX tend to be more mesenchymal, less epithelial, less NE but not more NNE; ATOH1 CDX are enriched for inner ear hair cell genes and ATOH1 direct targets (as detected with peaks within 10Kb from TSS).

## ATOH1 vs ASCL1 GO enrichment analysis

### FGSEA

```{r}

# import pathway lists by gene name - important: GAGE uses entrez ID while here we use gene names

pathways.hallmark <- gmtPathways("/data/cep/shumphrey/MSigDB/h.all.v7.1.symbols.gmt")
pathways.kegg <- gmtPathways("/data/cep/shumphrey/MSigDB/c2.cp.kegg.v7.1.symbols.gmt")
pathways.reactome <- gmtPathways("/data/cep/shumphrey/MSigDB/c2.cp.reactome.v7.1.symbols.gmt")
pathways.allGeneOntology <- gmtPathways("/data/cep/shumphrey/MSigDB/c5.all.v7.1.symbols.gmt")
pathways.biologicalProcess <- gmtPathways("/data/cep/shumphrey/MSigDB/c5.bp.v7.1.symbols.gmt")
pathways.cellComponents <- gmtPathways("/data/cep/shumphrey/MSigDB/c5.cc.v7.1.symbols.gmt")
pathways.molecularFunc <- gmtPathways("/data/cep/shumphrey/MSigDB/c5.mf.v7.1.symbols.gmt")
pathways.immunoSig <- gmtPathways("/data/cep/shumphrey/MSigDB/c7.all.v7.1.symbols.gmt")

allPathways.list <- list(hallmark = pathways.hallmark, 
                         kegg = pathways.kegg, 
                         reactome = pathways.reactome, 
                         allGeneOntology = pathways.allGeneOntology, 
                         biologicalProcess = pathways.biologicalProcess, 
                         cellComponents = pathways.cellComponents, 
                         molecularFunc = pathways.molecularFunc, 
                         immunoSig = pathways.immunoSig)

gseaDat_ATOH1vASCL1 <- resShrinkATOH1vsASCL1_annot %>%
            filter(complete.cases(.)) %>% 
            mutate(rank.score = sign(-log2FoldChange)*log10(padj)) %>%
            dplyr::select(gene_name, rank.score) %>%
            na.omit() %>%
            distinct() %>% 
            deframe() %>%
            sort()

barplot(sort(gseaDat_ATOH1vASCL1, decreasing = T))

allPathways.fgsea.list <- map(allPathways.list, function(x) 
                            fgsea(pathways = x, stats = gseaDat_ATOH1vASCL1,  minSize=15, maxSize = 500, nperm=1000))

allPathways.fgsea.sig.list <- map(allPathways.fgsea.list, function(x) as_tibble(x) %>% 
                                                                          arrange(desc(NES)) %>% 
                                                                          filter(padj < 0.05)  
                                                                          )

head(allPathways.fgsea.sig.list)

#plotEnrichment(pathways.hallmark[["HALLMARK_KRAS_SIGNALING_UP"]], gseaDat_ATOH1vASCL1) + labs(title="KRAS signaling")

```

### gProfiler

```{r fig.width=12, fig.height=12}

# ascl1 genes

downregulated <- resShrinkATOH1vsASCL1_annot %>%
  filter(log2FoldChange < -2 & padj < 0.05) %>%
  arrange(padj) %>%
  dplyr::select(gene_name, padj) %>%
  deframe()

#atoh1 genes

upregulated <- resShrinkATOH1vsASCL1_annot %>%
  filter(log2FoldChange > 2 & padj < 0.05) %>%
  arrange(padj) %>%
  dplyr::select(gene_name, padj) %>%
  deframe()

# run gprofiler2

gp_down <- gost(names(downregulated), organism = "hsapiens", ordered_query = T, as_short_link = F, evcodes = TRUE)

gp_up <- gost(names(upregulated), organism = "hsapiens", ordered_query = T, as_short_link = F, evcodes = TRUE)

gp_down$result
gp_up$result


BP <- gp_down$result %>% 
  filter(source == "GO:BP") %>% 
  mutate(p_value = p_value*-1) %>%
  rbind(filter(gp_up$result, gp_up$result$source == "GO:BP")) %>%
  arrange(p_value)

# sketel system development is duplicated
#BP$term_name[50] <- "skeletal system development."

#write_csv(BP, file = paste0(params$Output_data_CDX_Bulk, "GProfiler_ATOH1vASCL1_upInASCL1_CDX.csv"))

gp.results <- ggplot(data = BP, 
                     aes(x = factor(term_name, levels = pull(arrange(BP, desc(p_value)), term_name)), 
                         y = sign(p_value) * -log10(abs(p_value)))) +
  geom_bar(stat = "identity", colour  = "black", width = .8) + 
  geom_hline(yintercept = -log10(0.05), colour = 'red') +
  geom_hline(yintercept = +log10(0.05), colour = 'red') +
    theme_bw() +
  coord_flip() +
  theme(text = element_text(size = 18),
      	axis.text.y = element_text(size = 14),
        axis.text.x = element_text(size = 18)) +
  labs(y = ("-log10(p value)"), x = "GO Biological Process")
       
gp.results

table(downtargets_100K_p001$GeneSymbol %in% names(upregulated))

knitr::knit_exit()

```



# EXPLORATORY analysis without accounting for results from BETA


## Overlap DE genes, upregulated in ATOH1 CDX, and genes that are differentially bound in ChIPSeq and are transcriptionally active (H3K4me3 signature)

```{r eval=FALSE, include=FALSE}
# ---------------
# Note: DB_and_H3K4me3 file generated in ChIPseq script
# ---------------


DB_and_H3K4me3_genes <- unique(DB_and_H3K4me3$geneName)

print(paste0("Genes that are differentially bound by ATOH1 and have H3K4me3 signature:"))
length(DB_and_H3K4me3_genes)

print(paste0("Gene overlap between differentially upregulated genes in ATOH1 CDX vs differentially bound regions with H3K4me3 signature:")) 
table(DB_and_H3K4me3_genes %in% resATOH1_only_significant_anno$geneName)

GOI <- DB_genes_unique[which(DB_and_H3K4me3_genes %in% resATOH1_only_significant_anno$geneName)]

```

```{r, eval=FALSE, include=FALSE, fig.height=6, fig.width=8}

plotData <- subset(vsd.annot, gene_name %in% GOI)

rownames(plotData) <- NULL

plotData <- dplyr::select(plotData, "ensembl_gene_id", starts_with("CDX")) %>%
  column_to_rownames(var = "ensembl_gene_id")

plotHeatmap <- pheatmap((plotData),
                  color = PurOr_palette,
                  #border_color = "grey60",
                  treeheight_row = 15,
                  treeheight_col = 15,
                  cluster_rows = TRUE,
                  cluster_cols = TRUE,
                  dendrogram = "both",
                  scale = "row",
                  trace = "none",
                  cutree_cols = 3,
                  #fontsize_row = 14,
                  #cellwidth = 12,
                  #cellheight = 16,
                  show_rownames = FALSE,
                  annotation_col = colInfo.agg, 
                  annotation_colors = list(TF = c(ASCL1 = "#08519c", NEUROD1 = "#dd1c77", POU2F3 = "#31a354", ATOH1 = "#feb24c")),
                  main = "Unscaled heatmap for 285 genes upregulated in ATOH1 CDX and \n differentially bound in ATOH1 ChIPseq, with H3K4me3 signature")

```

It looks like these 285 genes which are upregulated in ATOH1 CDX, bound by ATOH1 in ChIPseq and having a H3K4me3 signature, are not very specific to ATOH1 CDX, being variably expressed across the CDX panel.
This is also evident in the heatmap above, where the genes at the bottom of the heatmap are shared between ATOH1 and NEUROD1 CDX.

*Side note: because ATOH1 and NEUROD1 CDX share expression of NEUROD1, is it possible that there's a big overlap between the two groups due to activity of NEUROD1?*

## What if ATOH1 both stimulates and represses expression of target genes?

Hypothesis: ATOH1 binding does not only mean gene expression, but also target gene repression i.e. Notch signalling pathway.

Here I will analyse DE genes in both ways, upregulated and downregulated in ATOH1 CDX.

```{r eval=FALSE, include=FALSE}

#resATOH1vsOther_significant_anno

GOI <- DB_genes_unique[which(DB_genes_unique %in% resATOH1vsOther_significant_anno$geneName)]

print(paste0("Gene overlap between differentially deregulated (up- or downregulated) genes in ATOH1 CDX vs differentially bound regions:")) 
table(DB_genes_unique %in% resATOH1vsOther_significant_anno$geneName)


```

```{r eval=FALSE, include=FALSE}

plotData <- subset(vsd.annot, gene_name %in% GOI)

rownames(plotData) <- NULL

plotData <- dplyr::select(plotData, "ensembl_gene_id", starts_with("CDX")) %>%
  column_to_rownames(var = "ensembl_gene_id")

plotHeatmap <- pheatmap((plotData),
                  color = YlGnBu,
                  #border_color = "grey60",
                  treeheight_row = 15,
                  treeheight_col = 15,
                  cluster_rows = TRUE,
                  cluster_cols = TRUE,
                  dendrogram = "both",
                  scale = "none",
                  trace = "none",
                  #fontsize_row = 14,
                  #cellwidth = 12,
                  #cellheight = 16,
                  show_rownames = FALSE,
                  annotation_col = colInfo.agg, 
                  annotation_colors = list(TF = c(ASCL1 = "#08519c", NEUROD1 = "#dd1c77", POU2F3 = "#31a354", ATOH1 = "#feb24c")),
                  main = "Unscaled heatmap for 1394 DE and DB genes")

plotHeatmap <- pheatmap((plotData),
                  color = PurOr_palette,
                  #border_color = "grey60",
                  treeheight_row = 15,
                  treeheight_col = 15,
                  cluster_rows = TRUE,
                  cluster_cols = TRUE,
                  dendrogram = "both",
                  scale = "row",
                  trace = "none",
                  #fontsize_row = 14,
                  #cellwidth = 12,
                  #cellheight = 16,
                  show_rownames = FALSE,
                  annotation_col = colInfo.agg, 
                  annotation_colors = list(TF = c(ASCL1 = "#08519c", NEUROD1 = "#dd1c77", POU2F3 = "#31a354", ATOH1 = "#feb24c")), 
                  main = "Scaled heatmap for 1394 DE and DB genes")

```